### PR TITLE
[Merged by Bors] - feat(CategoryTheory): Equivalence between certain comma categories and products

### PR DESCRIPTION
**Pull request by javra**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Comma/Basic.lean [hunk 1]
```diff
@@ -409,6 +411,37 @@ def post (L : A ⥤ T) (R : B ⥤ T) (F : T ⥤ C) : Comma L R ⥤ Comma (L ⋙
...
+is equivalent to their product. -/
+def PUnitPUnitIso (L : A ⥤ Discrete PUnit) (R : B ⥤ Discrete PUnit) :
+    Comma L R ≌ A × B :=
+  Equivalence.mk ((fst L R).prod' (snd L R)) (PUnitPUnitFromProd L R)
+    { hom := 𝟙 _, inv := 𝟙 _ }
+    { hom := 𝟙 _, inv := 𝟙 _ }
+
+/-- Taking the comma category of a functor into `A ⥤ Discrete PUnit` and the identity
+`Discrete PUnit ⥤ Discrete PUnit` results in a category equivalent to `A`. -/
+def PUnitIdIso (L : A ⥤ Discrete PUnit) : Comma L (Functor.id _) ≌ A :=
```
Comment (TwoFX) [inline Mathlib/CategoryTheory/Comma/Basic.lean:437]: Can you show that the forward direction of this is naturally isomorphic to `Comma.fst`?

File: Mathlib/CategoryTheory/Comma/Basic.lean [hunk 2]
```diff
@@ -409,6 +411,37 @@ def post (L : A ⥤ T) (R : B ⥤ T) (F : T ⥤ C) : Comma L R ⥤ Comma (L ⋙
       right := f.right
       w := by simp only [Functor.comp_map, ← F.map_comp, f.w] }
 
+/-- The canonical functor from the product of two categories to the comma category of their
+respective functors into `Discrete PUnit`. -/
+@[simps]
+def PUnitPUnitFromProd (L : A ⥤ Discrete PUnit) (R : B ⥤ Discrete PUnit) :
```
Comment (TwoFX) [inline Mathlib/CategoryTheory/Comma/Basic.lean:417]: I would call this one `fromProd` and the next one `equivProd`.

File: Mathlib/CategoryTheory/Comma/Basic.lean (hunk 1 above)
Comment (TwoFX) [inline Mathlib/CategoryTheory/Comma/Basic.lean:437]: `pUnitIdEquiv` maybe? Or `toPUnitIdEquiv`?

Comment (TwoFX) [reply Mathlib/CategoryTheory/Comma/Basic.lean:437]: Also, I guess this statement actually generalized to `Comma L R` for any `R : Discrete PUnit => Discrete PUnit` (which is of course the identity).

Comment (javra) [reply Mathlib/CategoryTheory/Comma/Basic.lean:437]: They're actually equal on the nose to `Comma.fst` and `Comma.snd`

**Review by TwoFX — APPROVED** [review]
Thanks!
bors r+

File: Mathlib/CategoryTheory/Comma/Basic.lean [hunk 3]
```diff
@@ -409,6 +410,54 @@ def post (L : A ⥤ T) (R : B ⥤ T) (F : T ⥤ C) : Comma L R ⥤ Comma (L ⋙
       right := f.right
       w := by simp only [Functor.comp_map, ← F.map_comp, f.w] }
 
+/-- The canonical functor from the product of two categories to the comma category of their
```
Comment (mattrobball) [inline Mathlib/CategoryTheory/Comma/Basic.lean:413]: This should be a split into a separate file and not included in `Basic`. 

See http://speed.lean-fro.org/mathlib4/run-detail/99a581a1-3912-4f3b-aeab-4f557e34d7f0 It almost triples the instructions and is used in a more targeted fashion that results of `Basic`.

Comment (javra) [reply Mathlib/CategoryTheory/Comma/Basic.lean:413]: We could put it into a `Discrete.lean` file you mean?

**Pull request closed without merging**